src/Microsoft.ML.Tokenizers/Model/Tiktoken.cs

```diff
@@ -0,0 +1,415 @@
+
+                            vocab[decodedToken] = rank;
+                        }
+                        else
+                        {
+                            throw new FormatException($"Can't parse {tokens[1]} to integer");
+                        }
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new InvalidOperationException($"Failed to load from BPE encoder file stream: {ex.Message}", ex);
+            }
+
+            return (encoder, vocab, decoder);
+        }
+
+        /// <summary>
+        /// Gets the dictionary mapping special tokens to Ids.
+        /// </summary>
+        /// <returns>The dictionary mapping special tokens to Ids.</returns>
+        public IReadOnlyDictionary<string, int>? SpecialTokensEncoder => _specialTokensEncoder;
+
+        /// <summary>
+        /// Tokenize a sequence string to a list of tokens.
+        /// </summary>
+        /// <param name="sequence">The sequence to tokenize.</param>
+        /// <returns>The list of tokens generated from the sequence tokenization.</returns>
+        public override IReadOnlyList<Token> Tokenize(string sequence) => Tokenize(sequence, isSpecialToken: false);
+
+        /// <summary>
+        /// Tokenize a split sequence string to a list of tokens.
+        /// </summary>
+        /// <param name="sequence">The text to tokenize.</param>
+        /// <param name="isSpecialToken">Indicate if the token is a special token.</param>
+        /// <returns>The list of tokens generated from the sequence tokenization.</returns>
+        public override IReadOnlyList<Token> Tokenize(string sequence, bool isSpecialToken)
+        {
+            List<Token> tokens;
+
+            if (string.IsNullOrEmpty(sequence))
+            {
+                return _emptyTokenList;
+            }
+
+            if (isSpecialToken)
+            {
+                if (_specialTokensEncoder is null)
+                {
+                    throw new InvalidOperationException($"The tokenizer doesn't have special tokens");
+                }
+
+                if (_specialTokensEncoder.TryGetValue(sequence, out int id))
+                {
+                    return new List<Token> { new(id, sequence, (0, sequence.Length)) };
+                }
+
+                throw new InvalidOperationException($"The special token {sequence} doesn't exist in the tokenizer");
+            }
+
+            if (_cache.Lookup(sequence, out int[] ids))
+            {
+                tokens = new(ids.Length);
+                tokens.Add(new Token(ids[0], sequence, (0, sequence.Length)));
+                for (int i = 1; i < ids.Length; i++)
+                {
+                    // One word split mapped to multiple Ids. Make the offset of the remaining token point at the end with zero width.
+                    tokens.Add(new Token(ids[i], "", (sequence.Length, sequence.Length)));
```

Comment:
I don't understand the Tokens being produced here. For any given string input, one token is returned that represents the entire sequence and then the rest all represent ""?

Reply:
I have a comment in the code (not in this specific spot) explaining the Tiktoken encoding is not one to one between tokens and Ids. When we encounter this case we return the whole token associated with the first Id and then return the rest of Ids mapped to empty string. The following comment has a good example of such case. 

```C#
                // Tiktoken does not ensure a one-to-one mapping between IDs and tokens. Consequently, decoding individual IDs into tokens is not supported;
                // instead, decoding all IDs must be done collectively.
                // Here is example of case that map one character to multiple Ids:
                // '⭐' U-2B50 is mapped to Ids [2928, 99834] in the Tiktoken model.
                // In other words, the character '⭐' has UTF-8 encoded bytes 0xE2, 0xAD, 0x90, Tiktoken will map 0xE2 to [2928] and 0xAD, 0x90 to [99834].
```